cedar-lean/Cedar/Thm/Partial/Evaluation.lean

```diff
@@ -0,0 +1,113 @@
+import Cedar.Thm.Partial.Evaluation.Unary
+import Cedar.Thm.Partial.Evaluation.Var
+import Cedar.Thm.Data.Control
+
+/-! This file contains theorems about Cedar's partial evaluator. -/
+
+namespace Cedar.Thm.Partial.Evaluation
+
+open Cedar.Data
+open Cedar.Partial (Unknown)
+open Cedar.Spec (Error Result)
+
+/--
+  Partial evaluation with concrete inputs gives the same output as
+  concrete evaluation with those inputs
+-/
+theorem on_concrete_eqv_concrete_eval {expr : Spec.Expr} {request : Spec.Request} {entities : Spec.Entities}
+  (wf : request.WellFormed) :
+  Partial.evaluate expr request entities = (Spec.evaluate expr request entities).map Partial.Value.value
+:= by
+  cases expr <;> simp only [Spec.Expr.asPartialExpr]
+  case lit p => simp [Partial.evaluate, Spec.evaluate, Except.map]
+  case var v => exact @Var.on_concrete_eqv_concrete_eval v request entities wf
+  case and x₁ x₂ =>
+    have ih₁ := @on_concrete_eqv_concrete_eval x₁ request entities wf
+    have ih₂ := @on_concrete_eqv_concrete_eval x₂ request entities wf
+    exact And.on_concrete_eqv_concrete_eval ih₁ ih₂
+  case or x₁ x₂ =>
+    have ih₁ := @on_concrete_eqv_concrete_eval x₁ request entities wf
+    have ih₂ := @on_concrete_eqv_concrete_eval x₂ request entities wf
+    exact Or.on_concrete_eqv_concrete_eval ih₁ ih₂
+  case ite x₁ x₂ x₃ =>
+    have ih₁ := @on_concrete_eqv_concrete_eval x₁ request entities wf
+    have ih₂ := @on_concrete_eqv_concrete_eval x₂ request entities wf
+    have ih₃ := @on_concrete_eqv_concrete_eval x₃ request entities wf
+    exact Ite.on_concrete_eqv_concrete_eval ih₁ ih₂ ih₃
+  case unaryApp op x₁ =>
+    have ih₁ := @on_concrete_eqv_concrete_eval x₁ request entities wf
+    exact Unary.on_concrete_eqv_concrete_eval ih₁
+  case binaryApp op x₁ x₂ =>
+    have ih₁ := @on_concrete_eqv_concrete_eval x₁ request entities wf
+    have ih₂ := @on_concrete_eqv_concrete_eval x₂ request entities wf
+    exact Binary.on_concrete_eqv_concrete_eval ih₁ ih₂
+  case getAttr x₁ attr =>
+    have ih₁ := @on_concrete_eqv_concrete_eval x₁ request entities wf
+    exact GetAttr.on_concrete_eqv_concrete_eval ih₁
+  case hasAttr x₁ attr =>
+    have ih₁ := @on_concrete_eqv_concrete_eval x₁ request entities wf
+    exact HasAttr.on_concrete_eqv_concrete_eval ih₁
+  case set xs =>
+    have ih : ∀ x ∈ xs, Partial.evaluate x request entities = (Spec.evaluate x request entities).map Partial.Value.value := by
+      intro x h₁
+      apply @on_concrete_eqv_concrete_eval x request entities wf
+    exact Set.on_concrete_eqv_concrete_eval ih
+  case record attrs =>
+    -- rw [List.map_attach₂ (λ x => (x.fst, Spec.Expr.asPartialExpr x.snd))]
+    have ih : ∀ kv ∈ attrs, Partial.evaluate kv.snd request entities = (Spec.evaluate kv.snd request entities).map Partial.Value.value := by
+      intro kv h₁
+      have h₂ : sizeOf kv.snd <= sizeOf kv := by simp only [sizeOf, Prod._sizeOf_1] ; omega
+      have : sizeOf kv.snd < sizeOf (Spec.Expr.record attrs) := by sorry
```

Comment:
I think you need an explicit `terminationBy` and then `decreasingBy` proof for this theorem.  

This will likely trigger two proof obligations:  one for sets and one for records.  But you'll be able to push it through.  I've seen this before too, where the addition of the missing proof obligation to the proof context is not enough.  The explicit proof will work in those cases.

Reply:
Thanks for the tip.  Termination proof is now passing.